cpp/src/parquet/encoding.h

```diff
@@ -317,6 +317,13 @@ class TypedDecoder : virtual public Decoder {
                           int64_t valid_bits_offset,
                           typename EncodingTraits<DType>::Accumulator* out) = 0;
 
+  virtual int DecodeArrow_opt(int num_values, int null_count, const uint8_t* valid_bits,
+                              int32_t* offset,
+                              std::shared_ptr<::arrow::ResizableBuffer>& values,
+                              int64_t valid_bits_offset, int32_t* bianry_length) {
+    return 0;
```

Comment:
Just  follow previous  DecodeArrow() function call stack usages.

Reply:
My gut feeling is still that they are not necessary. Though it might be wrong.

Reply:
It seems like we are adding this because the other methods are based on builders (in the accumulator), and builder don't provide a way to transfer multiple values in one `memcpy`. Does that sound right?

I wonder if we could add such a method on builders, and that might be a cleaner solution. Something like:

```cpp
class BaseBinaryBuilder {
...
  Status UnsafeAppendValues(const uint8_t* values, int64_t length, const uint8_t* valid_bytes = NULL_PTR) { ... }
};
```

The reason getting back to builders might be good is that I don't think these changes handle the case where there are more values than can fit into a StringArray. The current implementation will split it up into chunks if it reaches capacity, and I think we need to keep that behavior.

Reply:
Also, I'll say it again, but another possibility is to change the definition of `EncodingTraits<ByteArrayType>::Accumulator` from:
```c++
template <>
struct EncodingTraits<ByteArrayType> {
  using Encoder = ByteArrayEncoder;
  using Decoder = ByteArrayDecoder;

  /// \brief Internal helper class for decoding BYTE_ARRAY data where we can
  /// overflow the capacity of a single arrow::BinaryArray
  struct Accumulator {
    std::unique_ptr<::arrow::BinaryBuilder> builder;
    std::vector<std::shared_ptr<::arrow::Array>> chunks;
  };
  using ArrowType = ::arrow::BinaryType;
  using DictAccumulator = ::arrow::Dictionary32Builder<::arrow::BinaryType>;
};
```
to:
```c++
template <>
struct EncodingTraits<ByteArrayType> {
  using Encoder = ByteArrayEncoder;
  using Decoder = ByteArrayDecoder;

  /// \brief Internal helper class for decoding BYTE_ARRAY data where we can
  /// overflow the capacity of a single arrow::Int32Array
  struct Accumulator {
    std::unique_ptr<::arrow::Int32Builder> offsets_builder;
    std::unique_ptr<::arrow::BufferBuilder> data_builder;
    std::vector<std::shared_ptr<::arrow::Array>> chunks;
  };
  using ArrowType = ::arrow::BinaryType;
  using DictAccumulator = ::arrow::Dictionary32Builder<::arrow::BinaryType>;
};
```

Either this or @wjones127 's suggestion would be better than adding a specialized method, IMHO.

Reply:
Thanks Antoine. I guess I'm just catching to up your understanding 😄

Reply:
@wgtmac I'm not sure if you are objecting to my proposal above. A separate `BufferBuilder` for the string data allows to presize for the computed total length, so it should address your concern.

Reply:
> @wgtmac I'm not sure if you are objecting to my proposal above. A separate `BufferBuilder` for the string data allows to presize for the computed total length, so it should address your concern.

Sorry I didn't make it clear. I'm not objecting your proposal. Instead I just summarized another optimization we have done before if reusing `arrow::RecordBatch` on the same reader is possible. Internal experiment reveals that repeated buffer allocation and resize operation are non-negligible overhead, especially for wide columns (e.g. 1000+ columns).  @pitrou